primitives/src/crh/poseidon/batched_crh.rs

```diff
@@ -1,126 +1,99 @@
-extern crate rand;
-extern crate rayon;
-
-use algebra::{PrimeField, MulShort};
+use algebra::PrimeField;
 use std::marker::PhantomData;
 use crate::crh::BatchFieldBasedHash;
-use crate::{Error, PoseidonParameters, matrix_mix_short, PoseidonHash};
+use crate::{Error, PoseidonParameters, PoseidonHash, BatchSBox};
+use rayon::prelude::*;
 
-/// Batch evaluation for x^{-1}-Poseidon with MDS matrix supporting short Montgomery multiplication.
-pub struct PoseidonBatchHash<F: PrimeField, P: PoseidonParameters<Fr = F>>{
+pub struct PoseidonBatchHash<F: PrimeField, P: PoseidonParameters<Fr = F>, SB: BatchSBox<Field = F, Parameters = P>>
```

Comment:
Is it really worth to do introduce  an extra trait BatchSBox? 
I wouldn't know of any other than the inversion SBox.